### ISSUE TITLE
“indirection”

### ISSUE THREAD

**Issue by isomorphisms** [Raku/doc#2589]
## The problem
https://docs.perl6.org/language/containers

I don’t know what indirection is. And I don’t think objects are anything but a struct packaging nouns and verbs. Learning this wasted so much time and made me spend more of my life around programmers (which is a bad thing).

## Suggestions
Let’s forget object oriented religion or references to ruby. Pretend they never existed. Doing so creates dependencies of knowledge which themselves have dependencies, and which aren’t ultimately worth the time.

**Comment by JJ** [comment]
Can you please clarify what's the problem with "indirection"?

**Comment by b2gills** [comment]
Objects in Perl6 have [almost] nothing in common with structs.  
Thinking like that will most likely cause confusion.
(Structs don't have a meta object protocol)

I also don't know what Ruby has to do with anything.

“_indirect_” means the opposite of “_direct_”

Imagine if you wanted to give a message to Bob.  
If you talk directly to Bob there is no indirection.  
If you instead give Alice the message and tell her to tell Bob, there is a level of indirection.

“_indirection_” is only in the introductory statement, which only exists to make our docs be more conversational.
If that statement really confuses you, you can just skip it.

---

That page is about containers, which is a [mostly] separate topic to objects.

It is a **very** important topic when it comes to Perl6.
It might very well be the most important topic.

```perl6
# $foo is a scalar container
my $foo = 1..5;

for $foo { say $_ }
# 1..5

# decontainerize $foo
for $foo<> { say $_ }
# 1
# 2
# 3
# 4
# 5
```

By “_decontainerize_” I mean get the value out of the container and operate on it instead.

**Comment by pmurias** [comment]
To me the introduction is confusing. Knowing how containers work I can't easily explain what the "indirection involved in dealing with variables" is.

**Comment by AlexDaniel** [comment]
I think our docs can benefit from some [ELI5](https://en.wiktionary.org/wiki/ELI5)-like explanations.

> Imagine if you wanted to give a message to Bob.

I love it. In fact, maybe we should have a `ELI5` label here in this repo, and creating tickets like this one should be encouraged.

On the other hand, the page is subtitled “A low-level explanation of Perl 6 containers”. But where is the high-level explanation then?

**Comment by AlexDaniel** [comment]
OK, I created a new label… feel free to object or remove it.

**Comment by JJ** [comment]
There are few, if any, 5-year old programmers. Those who are would generally go for Scratch, not Perl 6. In general, I would find ELI5 language condescending, if not outright useless.

**Comment by jmaslak** [comment]
FWIW, I did program at age 5 (BASIC). I learned what I new from a college text that I copied examples from. Don't underestimate 5 year olds.  :)  But yes, ELI5 might be a bit much!

**Comment by rafaelschipiura** [comment]
ELI5 doesn't mean explaining something for literal 5 year olds. It's about explaining to a layman.
